// repomix/lib/core/metrics/tokenCountCache.js
import { createHash, randomBytes } from 'node:crypto';
import { existsSync as fsExistsSync } from 'node:fs';
import fs from 'node:fs/promises';
import path from 'node:path';
import { logger } from '../../shared/logger.js';
import { getRepomixTmpDir } from '../../shared/tmpDir.js';
const CACHE_VERSION = 1;
export const MAX_CACHE_ENTRIES = 100_000;
const CACHE_SUBDIR_NAME = 'cache';
const CACHE_FILE_NAME = 'token-counts.json';
const SEEN_SUBDIR_NAME = 'seen';
const createState = () => ({
    loaded: false,
    dirty: false,
    revision: 0,
    entries: new Map(),
});
let state = createState();
export const getCacheFilePath = () => {
    const override = process.env.REPOMIX_TOKEN_CACHE_PATH;
    if (override)
        return override;
    return path.join(getRepomixTmpDir(), CACHE_SUBDIR_NAME, CACHE_FILE_NAME);
};
export const isCacheDisabled = () => {
    return process.env.REPOMIX_TOKEN_CACHE === '0';
};
export const tokenCountCacheFileExistsSync = () => {
    if (isCacheDisabled())
        return false;
    try {
        return fsExistsSync(getCacheFilePath());
    }
    catch {
        return false;
    }
};
const seenMarkerKey = (rootDirs) => {
    const joined = rootDirs
        .map((d) => path.resolve(d))
        .sort()
        .join('\n');
    return createHash('md5').update(joined).digest('hex');
};
export const getRepoSeenMarkerPath = (rootDirs) => {
    const cacheDir = path.dirname(getCacheFilePath());
    return path.join(cacheDir, SEEN_SUBDIR_NAME, seenMarkerKey(rootDirs));
};
export const tokenCountCacheSeenMarkerExistsSync = (rootDirs) => {
    if (isCacheDisabled())
        return false;
    try {
        return fsExistsSync(getRepoSeenMarkerPath(rootDirs));
    }
    catch {
        return false;
    }
};
const markRepoSeen = async (rootDirs) => {
    if (isCacheDisabled() || rootDirs.length === 0)
        return;
    const markerPath = getRepoSeenMarkerPath(rootDirs);
    try {
        await fs.mkdir(path.dirname(markerPath), { recursive: true });
        await fs.writeFile(markerPath, '', { mode: 0o600 });
    }
    catch (error) {
        logger.trace('Failed to touch repo-seen marker:', error);
    }
};
export const loadTokenCountCache = async () => {
    if (state.loaded)
        return;
    state.loaded = true;
    if (isCacheDisabled()) {
        logger.trace('Token count cache disabled via REPOMIX_TOKEN_CACHE=0');
        return;
    }
    const cacheFile = getCacheFilePath();
    try {
        const raw = await fs.readFile(cacheFile, 'utf8');
        const data = JSON.parse(raw);
        if (data?.version !== CACHE_VERSION || !data.entries) {
            logger.trace('Token count cache version mismatch — discarding');
            return;
        }
        for (const key in data.entries) {
            const value = data.entries[key];
            if (typeof value === 'number') {
                state.entries.set(key, value);
            }
        }
        logger.trace(`Loaded ${state.entries.size} token count cache entries from ${cacheFile}`);
    }
    catch {
        logger.trace('Token count cache not found or unreadable — starting fresh');
    }
};
export const saveTokenCountCache = async (rootDirs = []) => {
    if (!isCacheDisabled() && rootDirs.length > 0 && tokenCountCacheFileExistsSync()) {
        await markRepoSeen(rootDirs);
    }
    if (!state.dirty || state.entries.size === 0)
        return;
    if (isCacheDisabled())
        return;
    const cacheFile = getCacheFilePath();
    const cacheDir = path.dirname(cacheFile);
    try {
        await fs.mkdir(cacheDir, { recursive: true });
        if (state.entries.size > MAX_CACHE_ENTRIES) {
            const toRemove = state.entries.size - MAX_CACHE_ENTRIES;
            const keys = state.entries.keys();
            for (let i = 0; i < toRemove; i++) {
                const oldest = keys.next().value;
                if (oldest === undefined)
                    break;
                state.entries.delete(oldest);
            }
        }
        const startRevision = state.revision;
        const data = {
            version: CACHE_VERSION,
            entries: Object.fromEntries(state.entries),
        };
        const uniqueSuffix = randomBytes(4).toString('hex');
        const tmpFile = `${cacheFile}.${process.pid}.${uniqueSuffix}.tmp`;
        await fs.writeFile(tmpFile, JSON.stringify(data), { mode: 0o600 });
        await fs.rename(tmpFile, cacheFile);
        if (state.revision === startRevision) {
            state.dirty = false;
        }
        else {
            state.dirty = true;
        }
        logger.trace(`Saved ${state.entries.size} token count cache entries to ${cacheFile}`);
        if (rootDirs.length > 0) {
            await markRepoSeen(rootDirs);
        }
    }
    catch (error) {
        logger.trace('Failed to save token count cache:', error);
    }
};
export const contentCacheKey = (encoding, content) => {
    const byteLength = Buffer.byteLength(content);
    const digest = createHash('md5').update(content).digest('hex').slice(0, 16);
    return `${encoding}:${byteLength}:${digest}`;
};
export const getCached = (key) => {
    if (isCacheDisabled())
        return undefined;
    return state.entries.get(key);
};
export const setCached = (key, tokenCount) => {
    if (isCacheDisabled())
        return;
    if (!state.entries.has(key) && state.entries.size >= MAX_CACHE_ENTRIES) {
        const oldest = state.entries.keys().next().value;
        if (oldest !== undefined) {
            state.entries.delete(oldest);
        }
    }
    state.entries.set(key, tokenCount);
    state.dirty = true;
    state.revision += 1;
};
export const __resetTokenCountCacheForTests = () => {
    state = createState();
};
